In [1]:
%pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
import datetime

In [3]:
df = pd.read_csv('Documents/Energy Data Analytics Career Plan/demanddata_2025.csv')

In [4]:
df.head()

,SETTLEMENT_DATE,SETTLEMENT_PERIOD,ND,TSD,ENGLAND_WALES_DEMAND,EMBEDDED_WIND_GENERATION,EMBEDDED_WIND_CAPACITY,EMBEDDED_SOLAR_GENERATION,EMBEDDED_SOLAR_CAPACITY,NON_BM_STOR,...,IFA_FLOW,IFA2_FLOW,BRITNED_FLOW,MOYLE_FLOW,EAST_WEST_FLOW,NEMO_FLOW,NSL_FLOW,ELECLINK_FLOW,VIKING_FLOW,GREENLINK_FLOW
0,2025-01-01,1,21036,26215,19918,5405,6606,0,19270,0,...,-1518,-1026,1,-232,-386,-558,96,0,241,0
1,2025-01-01,2,21222,26063,20167,5358,6606,0,19270,0,...,-1513,-1023,1,-124,-282,-564,51,0,262,0
2,2025-01-01,3,21385,25734,20328,5237,6606,0,19270,0,...,-1428,-922,1,-119,-384,-205,62,0,414,0
3,2025-01-01,4,20473,24988,19460,5115,6606,0,19270,0,...,-1423,-915,0,-119,-334,-193,37,0,423,0
4,2025-01-01,5,19634,24912,18654,5080,6606,0,19270,0,...,-1539,-899,0,-119,-363,-622,-282,0,288,0


In [5]:
df.shape # We expect (48) half hour settlement periods per day * (365) days in 2025 = 17520 rows 

(17520, 22)

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 17520 entries, 0 to 17519
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   SETTLEMENT_DATE            17520 non-null  str  
 1   SETTLEMENT_PERIOD          17520 non-null  int64
 2   ND                         17520 non-null  int64
 3   TSD                        17520 non-null  int64
 4   ENGLAND_WALES_DEMAND       17520 non-null  int64
 5   EMBEDDED_WIND_GENERATION   17520 non-null  int64
 6   EMBEDDED_WIND_CAPACITY     17520 non-null  int64
 7   EMBEDDED_SOLAR_GENERATION  17520 non-null  int64
 8   EMBEDDED_SOLAR_CAPACITY    17520 non-null  int64
 9   NON_BM_STOR                17520 non-null  int64
 10  PUMP_STORAGE_PUMPING       17520 non-null  int64
 11  SCOTTISH_TRANSFER          17520 non-null  int64
 12  IFA_FLOW                   17520 non-null  int64
 13  IFA2_FLOW                  17520 non-null  int64
 14  BRITNED_FLOW               17520 

In [7]:
df.columns

Index(['SETTLEMENT_DATE', 'SETTLEMENT_PERIOD', 'ND', 'TSD',
       'ENGLAND_WALES_DEMAND', 'EMBEDDED_WIND_GENERATION',
       'EMBEDDED_WIND_CAPACITY', 'EMBEDDED_SOLAR_GENERATION',
       'EMBEDDED_SOLAR_CAPACITY', 'NON_BM_STOR', 'PUMP_STORAGE_PUMPING',
       'SCOTTISH_TRANSFER', 'IFA_FLOW', 'IFA2_FLOW', 'BRITNED_FLOW',
       'MOYLE_FLOW', 'EAST_WEST_FLOW', 'NEMO_FLOW', 'NSL_FLOW',
       'ELECLINK_FLOW', 'VIKING_FLOW', 'GREENLINK_FLOW'],
      dtype='str')

In [8]:
df.isna().sum() # Examine missing values

SETTLEMENT_DATE              0
SETTLEMENT_PERIOD            0
ND                           0
TSD                          0
ENGLAND_WALES_DEMAND         0
EMBEDDED_WIND_GENERATION     0
EMBEDDED_WIND_CAPACITY       0
EMBEDDED_SOLAR_GENERATION    0
EMBEDDED_SOLAR_CAPACITY      0
NON_BM_STOR                  0
PUMP_STORAGE_PUMPING         0
SCOTTISH_TRANSFER            0
IFA_FLOW                     0
IFA2_FLOW                    0
BRITNED_FLOW                 0
MOYLE_FLOW                   0
EAST_WEST_FLOW               0
NEMO_FLOW                    0
NSL_FLOW                     0
ELECLINK_FLOW                0
VIKING_FLOW                  0
GREENLINK_FLOW               0
dtype: int64

In [9]:
df.duplicated().sum() # Check duplicates

np.int64(0)

In [10]:
df['SETTLEMENT_DATE'] = pd.to_datetime(df['SETTLEMENT_DATE'])
# Change data type from string to datetime

In [11]:
assert pd.api.types.is_datetime64_any_dtype(df["SETTLEMENT_DATE"]) # Check that 'settlement_date' is datetime dtype

In [12]:
df.dtypes

SETTLEMENT_DATE              datetime64[us]
SETTLEMENT_PERIOD                     int64
ND                                    int64
TSD                                   int64
ENGLAND_WALES_DEMAND                  int64
EMBEDDED_WIND_GENERATION              int64
EMBEDDED_WIND_CAPACITY                int64
EMBEDDED_SOLAR_GENERATION             int64
EMBEDDED_SOLAR_CAPACITY               int64
NON_BM_STOR                           int64
PUMP_STORAGE_PUMPING                  int64
SCOTTISH_TRANSFER                     int64
IFA_FLOW                              int64
IFA2_FLOW                             int64
BRITNED_FLOW                          int64
MOYLE_FLOW                            int64
EAST_WEST_FLOW                        int64
NEMO_FLOW                             int64
NSL_FLOW                              int64
ELECLINK_FLOW                         int64
VIKING_FLOW                           int64
GREENLINK_FLOW                        int64
dtype: object

In [13]:
df['ENGLAND_WALES_DEMAND'].min() # Sense check - is the answer negative?

np.int64(12040)

In [14]:
df['ENGLAND_WALES_DEMAND'].max() # Sense check - is the order of magnitude correct?

np.int64(41835)

In [15]:
# Question 1: Find average electricity demand in 2025

average_demand = df['ENGLAND_WALES_DEMAND'].mean()
average_demand = round(average_demand, 0)
print(f"Average GB electricity demand: {average_demand} MW")


Average GB electricity demand: 23996.0 MW


In [16]:
# Question 2: Find minimum and maximum demand, and when they occurred in 2025

min_demand = df.loc[df["ENGLAND_WALES_DEMAND"].idxmin(), ["SETTLEMENT_DATE", "SETTLEMENT_PERIOD", "ENGLAND_WALES_DEMAND"]]
print(min_demand)

max_demand = df.loc[df["ENGLAND_WALES_DEMAND"].idxmax(), ["SETTLEMENT_DATE", "SETTLEMENT_PERIOD", "ENGLAND_WALES_DEMAND"]]
print(max_demand)

SETTLEMENT_DATE         2025-05-25 00:00:00
SETTLEMENT_PERIOD                        30
ENGLAND_WALES_DEMAND                  12040
Name: 6939, dtype: object
SETTLEMENT_DATE         2025-01-09 00:00:00
SETTLEMENT_PERIOD                        35
ENGLAND_WALES_DEMAND                  41835
Name: 418, dtype: object


In [17]:
print(df[["SETTLEMENT_DATE", "SETTLEMENT_PERIOD"]].head()) 
# Settlement date is a datetime dtype. Output from question 2 tells me more data cleaning is required.
# Two problems arise:
# 1. "SETTLEMENT_DATE" returns the time at 00:00:00 for both minimum and maximum demand. 
# In the case of maximum demand this is nonsensical. I need to remove the time from the datetime object.
# 2. A new column, "SETTLEMENT_TIME" in the 24-hour clock needs to be created from "SETTLEMENT_PERIOD".
# Clock changes in March and October need to be accounted for.

  SETTLEMENT_DATE  SETTLEMENT_PERIOD
0      2025-01-01                  1
1      2025-01-01                  2
2      2025-01-01                  3
3      2025-01-01                  4
4      2025-01-01                  5


In [18]:
# Hardcoded 2025 Clock Change Logic - Solving problem 2
def calculate_2025_time(row):
    # Convert date to string format for easy matching
    date_str = row["SETTLEMENT_DATE"].strftime("%Y-%m-%d")
    period = row["SETTLEMENT_PERIOD"]

    # --- SPRING CLOCK CHANGE: 30 March 2025 (46 Periods) ---
    if date_str == "2025-03-30":
        if period > 2:
            # Skip the lost hour (add 2 extra periods worth of minutes)
            minutes_elapsed = (period + 1) * 30
        else:
            minutes_elapsed = (period - 1) * 30

    # --- AUTUMN CLOCK CHANGE: 26 October 2025 (50 Periods) ---
    elif date_str == "2025-10-26":
        if period > 6:
            # Repeat the hour (subtract 2 periods worth of minutes)
            minutes_elapsed = (period - 3) * 30
        else:
            minutes_elapsed = (period - 1) * 30

    # --- STANDARD DAY (48 Periods) ---
    else:
        minutes_elapsed = (period - 1) * 30

    # Calculate hours and minutes mathematically
    hours = (minutes_elapsed // 60) % 24
    minutes = minutes_elapsed % 60

    return f"{hours:02d}:{minutes:02d}"


# Apply the 2025 function using lambda
df["SETTLEMENT_TIME"] = df.apply(lambda row: calculate_2025_time(row), axis=1)

# Print sample dataframe with affected days and times to check the function and one non-affected day as control variable
target_dates = pd.to_datetime(["2025-03-30", "2025-10-26", "2025-10-27"])

print(
    df.loc[
       ((df["SETTLEMENT_DATE"].isin(target_dates)) & (
           df["SETTLEMENT_PERIOD"] <= 7)),
        ["SETTLEMENT_DATE", "SETTLEMENT_TIME", "SETTLEMENT_PERIOD"],
    ]
)




      SETTLEMENT_DATE SETTLEMENT_TIME  SETTLEMENT_PERIOD
4224       2025-03-30           00:00                  1
4225       2025-03-30           00:30                  2
4226       2025-03-30           02:00                  3
4227       2025-03-30           02:30                  4
4228       2025-03-30           03:00                  5
4229       2025-03-30           03:30                  6
4230       2025-03-30           04:00                  7
14302      2025-10-26           00:00                  1
14303      2025-10-26           00:30                  2
14304      2025-10-26           01:00                  3
14305      2025-10-26           01:30                  4
14306      2025-10-26           02:00                  5
14307      2025-10-26           02:30                  6
14308      2025-10-26           02:00                  7
14352      2025-10-27           00:00                  1
14353      2025-10-27           00:30                  2
14354      2025-10-27          

In [19]:
# Re-answer Question 2: Find minimum and maximum demand, and when they occurred in 2025

min_demand = df.loc[df["ENGLAND_WALES_DEMAND"].idxmin(), ["SETTLEMENT_DATE", "SETTLEMENT_TIME", "SETTLEMENT_PERIOD", "ENGLAND_WALES_DEMAND"]]
print(min_demand)

max_demand = df.loc[df["ENGLAND_WALES_DEMAND"].idxmax(), ["SETTLEMENT_DATE", "SETTLEMENT_TIME", "SETTLEMENT_PERIOD", "ENGLAND_WALES_DEMAND"]]
print(max_demand)

SETTLEMENT_DATE         2025-05-25 00:00:00
SETTLEMENT_TIME                       14:30
SETTLEMENT_PERIOD                        30
ENGLAND_WALES_DEMAND                  12040
Name: 6939, dtype: object
SETTLEMENT_DATE         2025-01-09 00:00:00
SETTLEMENT_TIME                       17:00
SETTLEMENT_PERIOD                        35
ENGLAND_WALES_DEMAND                  41835
Name: 418, dtype: object


The minimum electricity demand in GB in 2025 was 12,040 MW, which occurred on 25/5/2025 at 14:30. This was a Sunday in late spring. Surprisingly the lowest demand all year wasn't during night hours.
Maximum GB demand for 2025 was 41,835 MW. This occurred on 9/1/2025 at 17:00, a weekday during a dark, cold winter.

In [41]:
# Question 3 — Find average, minimum and maximum demand by month

import calendar

# Get a clean list of all 12 month names ['January', 'February', ...]
ordered_months = list(calendar.month_name)[1:]

# Extract month name and convert it into a ordered Categorical column
df["month_name"] = pd.Categorical(
    df["SETTLEMENT_DATE"].dt.month_name(),
    categories=ordered_months,
    ordered=True,
)

# Now, any groupby or sort operation will naturally stay in perfect calendar order!
monthly_demand_summary = (
    df.groupby("month_name", observed=False)["ENGLAND_WALES_DEMAND"]
    .agg(["mean", "min", "max"])
    .reset_index()
)

# Round the mean column to the nearest integer
monthly_demand_summary["mean"] = monthly_demand_summary["mean"].round(0).astype(int)

print(monthly_demand_summary)



   month_name   mean    min    max
0     January  30484  16289  41835
1    February  29166  17764  40133
2       March  25191  13056  36581
3       April  21378  12119  29911
4         May  19782  12040  27165
5        June  20144  13352  27291
6        July  20756  13908  26900
7      August  19988  12093  27115
8   September  21734  14804  30729
9     October  24443  14720  34652
10   November  27392  16199  39688
11   December  27827  17852  38612


This table shows demand peaks in winter and is lowest in summer, as expected. The month of minimum demand
is May, technically late spring, so this is a surprising result.

In [33]:
# Question 4: What are the highest-demand periods of 2025?

top_demand = df.sort_values("ENGLAND_WALES_DEMAND", ascending=False)

# What day of week correspond to the highest demand days?
# Create a new column with the day names
df["SETTLEMENT_DAY"] = df["SETTLEMENT_DATE"].dt.day_name()


print(top_demand[['SETTLEMENT_DAY', 'SETTLEMENT_DATE', 'SETTLEMENT_TIME', 'ENGLAND_WALES_DEMAND']].head(10))

    SETTLEMENT_DAY SETTLEMENT_DATE SETTLEMENT_TIME  ENGLAND_WALES_DEMAND
418       Thursday      2025-01-09           17:00                 41835
419       Thursday      2025-01-09           17:30                 41811
371      Wednesday      2025-01-08           17:30                 41620
370      Wednesday      2025-01-08           17:00                 41614
420       Thursday      2025-01-09           18:00                 41479
417       Thursday      2025-01-09           16:30                 41428
372      Wednesday      2025-01-08           18:00                 41389
369      Wednesday      2025-01-08           16:30                 41384
421       Thursday      2025-01-09           18:30                 41295
323        Tuesday      2025-01-07           17:30                 41182


The above table shows that the top 10 observations of highest electricity demand by settlement period occur 
in a three-day window in early January, from Tuesday 7/1/2025 to Thursday 9/1/2025. Settlement times for highest demand 
occur between 16:30 and 18:30 corresponding to usual work finish time, when there will be a surge in household heating and 
lighting demand as people arrive from work to a cold dark home in winter. 

In [38]:
# Question 5 — How variable is demand?

df["ENGLAND_WALES_DEMAND"].describe().round(0).astype(int).drop('count')

mean    23996
std      5651
min     12040
25%     19638
50%     22904
75%     27386
max     41835
Name: ENGLAND_WALES_DEMAND, dtype: int64

In [40]:
demand_range = df["ENGLAND_WALES_DEMAND"].max() - df["ENGLAND_WALES_DEMAND"].min()
print(demand_range)

29795


A summary statistics table for electricity demand during 2025 is given above. The range of demand throughout
the year 2025 is 29,795 MW.